starter_project/autonomy/src/perception.cpp

```diff
@@ -54,15 +61,60 @@ namespace mrover {
         tags.clear(); // Clear old tags in output vector
 
         // TODO: implement me!
+        // cv::aruco::ArucoDetector detector = cv::aruco::ArucoDetector(mTagDictionary, mTagDetectorParams);
+
+        cv::aruco::detectMarkers(image, mTagDictionary, mTagCorners, mTagIds, mTagDetectorParams);
+
+        std::vector<float> closeness_metrics;
+        std::vector<std::pair<float, float>> centers;
+
+        for (std::vector<cv::Point2f> &corner_set : mTagCorners) {
+            closeness_metrics.push_back(getClosenessMetricFromTagCorners(image, corner_set));
+            centers.push_back(getCenterFromTagCorners(corner_set));
+        }
+
+        for (size_t i = 0; i < closeness_metrics.size(); i++) {
+            StarterProjectTag tag;
+
+            tag.tagId = mTagIds[i];
+            tag.xTagCenterPixel = centers[i].first;
+            tag.yTagCenterPixel = centers[i].second;
+            tag.closenessMetric = closeness_metrics[i];
+
+            tags.push_back(tag);
+        }
+
+
     }
 
-    StarterProjectTag Perception::selectTag(std::vector<StarterProjectTag> const& tags) { // NOLINT(*-convert-member-functions-to-static)
-        // TODO: implement me!
-        return {};
+    // double Perception::get_center_offset_magnitude_sq(float x, float y) const {
+    //     double x_offset = x - (double(Perception::image_width) / 2.0);
+    //     double y_offset = y - (double(Perception::image_height) / 2.0);
+
+    //     return x_offset*x_offset + y_offset*y_offset;
+    // }
+
+    StarterProjectTag Perception::selectTag(std::vector<StarterProjectTag> const& tags) const { // NOLINT(*-convert-member-functions-to-static)
+        if (tags.empty())
+            return {};
+
+        StarterProjectTag best = tags[0];
+        double best_dist = best.xTagCenterPixel*best.xTagCenterPixel + best.yTagCenterPixel*best.yTagCenterPixel;
+
+        for (int i = 1; i < tags.size(); i++) {
```

Comment:
Consider using https://en.cppreference.com/w/cpp/algorithm/ranges/min_element with a projection function

Also `tags.size()` returns a `size_t` so `int i` is narrowing